ssh-keyscan -p "443" "ssh.github.com" failed

Running `go test ./...` on a fresh clone at HEAD yields,
```
2022/12/24 14:17:08 Datadog Tracer v1.43.1 ERROR: Loading features: Get "http://localhost:8126/info": dial tcp [::1]:8126: connect: connection refused, 1 additional messages skipped (first occurrence: 24 Dec 22 14:17 PST)
2022/12/24 14:17:08 Datadog Tracer v1.43.1 ERROR: lost 1 traces: Post "http://localhost:8126/v0.4/traces": dial tcp [::1]:8126: connect: connection refused (occurred: 24 Dec 22 14:17 PST)
2022/12/24 14:17:08 warning: cannot parse BUILDKITE_CLEAN_CHECKOUT=blarg as bool, ignoring
2022/12/24 14:17:08 warning: cannot parse BUILDKITE_PLUGINS_ALWAYS_CLONE_FRESH=grarg as bool, ignoring
--- FAIL: TestAddingToKnownHosts (17.86s)
    --- FAIL: TestAddingToKnownHosts/ssh_url_with_port (15.27s)
        knownhosts_test.go:78: kh.AddFromRespository("ssh://git@ssh.github.com:443/var/cache/git/project.git") = Failed to add "ssh.github.com:443" to known_hosts file "ssh://git@ssh.github.com:443/var/cache/git/project.git": Could not  `ssh-keyscan`: `ssh-keyscan -p "443" "ssh.github.com"` failed
        knownhosts_test.go:86: kh.Contains("ssh.github.com:443") = false, want true
FAIL
FAIL    github.com/buildkite/agent/v3/bootstrap 17.877s
--- FAIL: TestCheckingOutLocalGitProjectWithSubmodulesDisabled (0.09s)
    checkout_integration_test.go:291: tester.Repo.Execute(submodule, add, "/tmp/git-repo271476945") error = exit status 128
        out = Cloning into '/tmp/git-repo989430050/git-repo271476945'...
        fatal: transport 'file' not allowed
        fatal: clone of '/tmp/git-repo271476945' into submodule path '/tmp/git-repo989430050/git-repo271476945' failed
--- FAIL: TestCheckingOutLocalGitProjectWithSubmodules (0.11s)
    checkout_integration_test.go:208: tester.Repo.Execute(submodule, add, "/tmp/git-repo891267893") error = exit status 128
        out = Cloning into '/tmp/git-repo2915220035/git-repo891267893'...
        fatal: transport 'file' not allowed
        fatal: clone of '/tmp/git-repo891267893' into submodule path '/tmp/git-repo2915220035/git-repo891267893' failed
2022/12/24 14:17:19 Killing pid 54806
FAIL
FAIL    github.com/buildkite/agent/v3/bootstrap/integration     12.419s
```
___
Have been able to narrow it down to having run the tests while connected to a VPN, so not blocking. Thought it'd be at least worth documenting.

--- a/bootstrap/knownhosts_test.go
+++ b/bootstrap/knownhosts_test.go
@@ -3,88 +3,73 @@ package bootstrap
 import (
 	"context"
 	"fmt"
+	"net"
 	"os"
-	"path/filepath"
 	"testing"
 
 	"github.com/buildkite/agent/v3/bootstrap/shell"
-	"github.com/buildkite/bintest/v3"
+	"github.com/gliderlabs/ssh"
 )
 
 func TestAddingToKnownHosts(t *testing.T) {
 	t.Parallel()
 
-	var testCases = []struct {
-		Name       string
-		Repository string
-		Alias      string
-		Host       string
-	}{
-		{"git url", "git@github.com:buildkite/agent.git", "github.com", "github.com"},
-		{"git url with alias", "git@github.com-alias1:buildkite/agent.git", "github.com-alias1", "github.com"},
-		{"ssh url with port", "ssh://git@ssh.github.com:443/var/cache/git/project.git", "ssh.github.com:443", "ssh.github.com:443"},
+	// Start a fake SSH server for ssh-keyscan to poke at.
+	// It will generate its own host key.
+	// This uses ':0' to pick an ephemeral port to listen on.
+	svr := &ssh.Server{
+		Addr:    "localhost:0",
+		Version: "Fake SSH Server v0.1",
+	}
+	ln, err := net.Listen("tcp", svr.Addr)
+	if err != nil {
+		t.Fatalf("net.Listen(tcp, %q) error = %v", svr.Addr, err)
+	}
+	go svr.Serve(ln)
+	defer svr.Close()
+
+	hostAddr := ln.Addr().String()
+	repoURL := fmt.Sprintf("ssh://git@%s/var/cache/git/project.git", hostAddr)
+	t.Logf("Fake SSH server listening at address %s", hostAddr)
+
+	// Create a new empty known-hosts file to add to.
+	f, err := os.CreateTemp("", "known-hosts")
+	if err != nil {
+		t.Fatalf(`os.CreateTemp("", "known-hosts") error = %v`, err)
+	}
+	defer os.RemoveAll(f.Name())
+	if err := f.Close(); err != nil {
+		t.Fatalf("f.Close() = %v", err)
 	}
 
-	for _, tc := range testCases {
-		t.Run(tc.Name, func(t *testing.T) {
-			sh := shell.NewTestShell(t)
+	sh := shell.NewTestShell(t)
+	sh.Env.Set("PATH", os.Getenv("PATH"))
+	kh := knownHosts{
+		Shell: sh,
+		Path:  f.Name(),
+	}
 
-			ssh, err := bintest.NewMock("ssh")
-			if err != nil {
-				t.Fatalf("bintest.NewMock(ssh) error = %v", err)
-			}
-			defer ssh.CheckAndClose(t)
+	// The file should not contain this host
+	exists, err := kh.Contains(hostAddr)
+	if err != nil {
+		t.Errorf("kh.Contains(%q) error = %v", hostAddr, err)
+	}
+	if got, want := exists, false; got != want {
+		t.Errorf("kh.Contains(%q) = %t, want %t", hostAddr, got, want)
+	}
 
-			path := fmt.Sprintf("%s%c%s", filepath.Dir(ssh.Path), os.PathListSeparator, os.Getenv("PATH"))
+	// Add the host - this resolves any SSH client configuration
+	// (in case the URL contains an alias), and runs ssh-keyscan to get its key.
+	if err := kh.AddFromRepository(context.Background(), repoURL); err != nil {
+		t.Errorf("kh.AddFromRespository(%q) = %v", repoURL, err)
+	}
 
-			sh.Env.Set("PATH", path)
-
-			ssh.
-				Expect("-G", tc.Alias).
-				AndWriteToStderr(`unknown option -- G
-usage: ssh [-1246AaCfgKkMNnqsTtVvXxYy] [-b bind_address] [-c cipher_spec]
-           [-D [bind_address:]port] [-E log_file] [-e escape_char]
-           [-F configfile] [-I pkcs11] [-i identity_file]
-           [-L [bind_address:]port:host:hostport] [-l login_name] [-m mac_spec]
-           [-O ctl_cmd] [-o option] [-p port]
-           [-Q cipher | cipher-auth | mac | kex | key]
-           [-R [bind_address:]port:host:hostport] [-S ctl_path] [-W host:port]
-           [-w local_tun[:remote_tun]] [user@]hostname [command]`).
-				AndExitWith(255)
-
-			f, err := os.CreateTemp("", "known-hosts")
-			if err != nil {
-				t.Fatalf(`os.CreateTemp("", "known-hosts") error = %v`, err)
-			}
-			defer os.RemoveAll(f.Name())
-			if err := f.Close(); err != nil {
-				t.Fatalf("f.Close() = %v", err)
-			}
-
-			kh := knownHosts{
-				Shell: sh,
-				Path:  f.Name(),
-			}
-
-			exists, err := kh.Contains(tc.Host)
-			if err != nil {
-				t.Errorf("kh.Contains(%q) error = %v", tc.Host, err)
-			}
-			if got, want := exists, false; got != want {
-				t.Errorf("kh.Contains(%q) = %t, want %t", tc.Host, got, want)
-			}
-
-			if err := kh.AddFromRepository(context.Background(), tc.Repository); err != nil {
-				t.Errorf("kh.AddFromRespository(%q) = %v", tc.Repository, err)
-			}
-
-			exists, err = kh.Contains(tc.Host)
-			if err != nil {
-				t.Errorf("kh.Contains(%q) error = %v", tc.Host, err)
-			}
-			if got, want := exists, true; got != want {
-				t.Errorf("kh.Contains(%q) = %t, want %t", tc.Host, got, want)
-			}
-		})
+	// The file should now contain the host key
+	exists, err = kh.Contains(hostAddr)
+	if err != nil {
+		t.Errorf("kh.Contains(%q) error = %v", hostAddr, err)
+	}
+	if got, want := exists, true; got != want {
+		t.Errorf("kh.Contains(%q) = %t, want %t", hostAddr, got, want)
 	}
 }

--- a/go.mod
+++ b/go.mod
@@ -54,12 +54,14 @@ require (
 	github.com/DataDog/datadog-go v4.8.2+incompatible // indirect
 	github.com/DataDog/sketches-go v1.2.1 // indirect
 	github.com/Microsoft/go-winio v0.5.1 // indirect
+	github.com/anmitsu/go-shlex v0.0.0-20200514113438-38f4b401e2be // indirect
 	github.com/cenkalti/backoff/v4 v4.2.0 // indirect
 	github.com/cespare/xxhash/v2 v2.1.2 // indirect
 	github.com/cpuguy83/go-md2man/v2 v2.0.0-20190314233015-f79a8a8ca69d // indirect
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/dgraph-io/ristretto v0.1.0 // indirect
 	github.com/dustin/go-humanize v1.0.0 // indirect
+	github.com/gliderlabs/ssh v0.3.5 // indirect
 	github.com/go-logr/logr v1.2.3 // indirect
 	github.com/go-logr/stdr v1.2.2 // indirect
 	github.com/golang/glog v1.0.0 // indirect

